influxdb3_write/src/cache.rs

```diff
@@ -0,0 +1,393 @@
+use crate::persister::serialize_to_parquet;
+use crate::persister::Error;
+use crate::ParquetFile;
+use bytes::Bytes;
+use datafusion::execution::memory_pool::MemoryPool;
+use datafusion::physical_plan::SendableRecordBatchStream;
+use object_store::memory::InMemory;
+use object_store::path::Path as ObjPath;
+use object_store::ObjectStore;
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+type MetaData = RwLock<HashMap<String, HashMap<String, HashMap<String, ParquetFile>>>>;
+
+#[derive(Debug)]
+pub struct ParquetCache {
+    object_store: Arc<dyn ObjectStore>,
+    meta_data: MetaData,
+    mem_pool: Arc<dyn MemoryPool>,
+}
+
+impl ParquetCache {
+    /// Create a new `ParquetCache`
+    pub fn new(mem_pool: &Arc<dyn MemoryPool>) -> Self {
+        Self {
+            object_store: Arc::new(InMemory::new()),
+            meta_data: RwLock::new(HashMap::new()),
+            mem_pool: Arc::clone(mem_pool),
+        }
+    }
+
+    /// Get the parquet file metadata for a given database and table
+    pub fn get_parquet_files(&self, database_name: &str, table_name: &str) -> Vec<ParquetFile> {
+        self.meta_data
+            .read()
+            .get(database_name)
+            .and_then(|db| db.get(table_name))
+            .cloned()
+            .unwrap_or_default()
+            .into_values()
+            .collect()
+    }
+
+    /// Persist a new parquet file to the cache or pass an object store path to update a currently
+    /// existing file in the cache
+    // Note we want to hold across await points until everything is cleared
+    // before letting other tasks access the data
+    #[allow(clippy::await_holding_lock)]
+    pub async fn persist_parquet_file(
+        &self,
+        db_name: &str,
+        table_name: &str,
+        min_time: i64,
+        max_time: i64,
+        record_batches: SendableRecordBatchStream,
+        path: Option<ObjPath>,
+    ) -> Result<(), Error> {
+        // Lock the data structure until everything is written into the object
+        // store and metadata
+        let mut meta_data_lock = self.meta_data.write();
+        let parquet = serialize_to_parquet(Arc::clone(&self.mem_pool), record_batches).await?;
+        // Generate a path for this
+        let id = uuid::Uuid::new_v4();
+        let parquet_path =
+            path.unwrap_or_else(|| ObjPath::from(format!("{db_name}-{table_name}-{id}")));
+        let size_bytes = parquet.bytes.len() as u64;
+        let meta_data = parquet.meta_data;
+
+        self.object_store.put(&parquet_path, parquet.bytes).await?;
+
+        let path = parquet_path.to_string();
+        let parquet_files = || -> HashMap<String, ParquetFile> {
```

Comment:
Does this always get used? Looks like it only gets used if things aren't in certain maps, which makes me think we should only construct it if we need it.

Reply:
I can just inline it then!